src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,454 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Result;
+use std::{str, f64};
+use std::collections::BTreeMap;
+use std::result::Result as SResult;
+use serde_json::{self, Value};
+use std::io::Write;
+use util::codec::number::{NumberDecoder, NumberEncoder};
+use serde::ser::{Serialize, Serializer, SerializeTuple, SerializeMap};
+
+const TYPE_CODE_OBJECT: u8 = 0x01;
+const TYPE_CODE_ARRAY: u8 = 0x03;
+const TYPE_CODE_LITERAL: u8 = 0x04;
+const TYPE_CODE_I64: u8 = 0x09;
+const TYPE_CODE_DOUBLE: u8 = 0x0b;
+const TYPE_CODE_STRING: u8 = 0x0c;
+
+const JSON_LITERAL_NIL: u8 = 0x00;
+const JSON_LITERAL_TRUE: u8 = 0x01;
+const JSON_LITERAL_FALSE: u8 = 0x02;
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+const ERR_EMPTY_DOCUMENT: &str = "The document is empty";
+
+// Json implement type json used in tikv, it specifies the following
+// implementations:
+// 1. Serialize `json` values into binary representation, and reading values
+//  back from the binary representation.
+// 2. Serialize `json` values into readable string representation, and reading
+// values back from string representation.
+// The binary Json format from MySQL 5.7 is in the following link:
+// https://github.com/mysql/mysql-server/blob/5.7/sql/json_binary.h#L52
+// The only difference is that we use large `object` or large `array` for
+// the small corresponding ones. That means in our implementation there
+// is no difference between small `object` and big `object`, so does `array`.
+#[derive(Clone,PartialEq,Debug)]
+pub enum Json {
+    Object(BTreeMap<String, Json>),
+    Array(Vec<Json>),
+    Literal(u8),
+    I64(i64),
+    Double(f64),
+    String(String),
+}
+
+#[allow(dead_code)]
+impl Json {
+    pub fn deserialize(data: &[u8]) -> Result<Json> {
+        if data.is_empty() {
+            return Err(invalid_type!(ERR_EMPTY_DOCUMENT));
+        }
+        Json::decode(data[0], &data[LENGTH_TYPE..])
+    }
+
+    pub fn parse_from_string(s: &str) -> Result<Json> {
```

Comment:
we have two way to create a Json from string:
1) using that string as a primitive value,  e.g. `"hello", "1", ...`
2) parse from that string, e.g. `"\"hello\"", "[1, 2, 3]", ...`

we can implement (1) by `impl From<String> for Json`, so  we get `Json::from("abcde)`.
But which style is best for (2)? `from_str` or `parse_XXX`?